### PR TITLE
JS cleanup

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
The goal here is just to improve the source code a bit. I recommend to review one commit at a time, otherwise it might not make much sense. :)

The biggest commit is the second one: to prevent to have "global" variables declared in `main.js` (and thus prevent name conflict or overwriting), I moved such code into anonymous functions.

r? @kinnison 

cc @rust-lang/rustdoc

### CODE REVIEW

File: src/librustdoc/html/static/main.js [hunk 1]
```diff
@@ -465,27 +477,29 @@ function getSearchElement() {
...
+                    if (i < len - 1) {
+                        stripped = match[0] + stripped;
+                    }
+                    url = url.substring(0, url.length - match[0].length);
                 }
-                url = url.substring(0, url.length - match[0].length);
-            }
 
-            url += "/" + document.getElementsByClassName("version-selector")[0].value + stripped;
+                url += "/" + document.getElementsByClassName("version-selector")[0].value + stripped;
```
Comment (kinnison) [inline src/librustdoc/html/static/main.js:497]: I think tidy dislikes this line.

File: src/librustdoc/html/static/main.js [hunk 2]
```diff
@@ -450,7 +450,7 @@ function defocusSearchBar() {
                 set_fragment(cur_line_id);
             }
         }
-    })();
+    }());
```
Comment (kinnison) [inline src/librustdoc/html/static/main.js:454]: While this isn't bad, it seems spurious in a "move code" commit rather than a cleanup one.  You can just resolve this if you don't care, otherwise move it to a different commit when you next rebase.

**Review by kinnison — COMMENTED** [review]
Fundamentally this looks like a reasonable cleanup.

I'm not sure of the value of the `window.FooBar` over `function FooBar` but I'm assuming you know more than I.  The lint fixes look right, but again I'm not the best at JS.

r=me modulo the tweaks

File: src/librustdoc/html/static/storage.js [hunk 3]
```diff
@@ -27,14 +27,15 @@ function removeClass(elem, className) {
 function onEach(arr, func, reversed) {
     if (arr && arr.length > 0 && func) {
         var length = arr.length;
+        var i;
```
Comment (hbina) [inline src/librustdoc/html/static/storage.js:30]: Is this really necessary? Each `i` is local to the loop so declaring it out here makes reasoning the loop harder.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/storage.js:30]: JS `var`s don't seem to understand the principle of scope. :)

Comment (bjorn3) [reply src/librustdoc/html/static/storage.js:30]: In JS a `var` is hoisted to the top of the function. The new `let` and `const` don't have this behavior.

Comment (workingjubilee) [reply src/librustdoc/html/static/storage.js:30]: re: fold, yes, browsers implemented Array.prototype.reduce() a long time ago.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce

Much like `Iter` and `for` loops are equivalent in Rust, `reduce` and `map` are equivalent to `for` loops in JS for similar reasons.
